extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/QuarkusWorkerPoolRegistry.java

```diff
@@ -73,6 +149,20 @@ public <T> Uni<T> executeWork(Context currentContext, Uni<T> uni, String workerN
         }
     }
 
+    private <T> Uni<T> runOnVirtualThread(Context currentContext, Uni<T> uni, Semaphore semaphore) {
+        return uni.runSubscriptionOn(command -> VIRTUAL_EXECUTOR_SUPPLIER.get().execute(command))
+                .onItemOrFailure().transformToUni((item, failure) -> {
+                    semaphore.release();
```

Comment:
What about cancellation? 

I think, onTermination might be better suited in this case.

CC @jponge

Reply:
Yes, `onTermination` covers all final states.

Reply:
can't transform a cancellation to a uni and perform the same logic as in onItemOrFailure so I just added a handler for cancellation that release the semaphore

Reply:
You need onCancelleation to release the semaphore.

The rest does not change